tests/MongoDB.Driver.Core.Tests/Specifications/server-discovery-and-monitoring/TestRunner.cs

```diff
@@ -57,18 +62,101 @@ public void RunTestDefinition(JsonDrivenTestCase testCase)
             }
         }
 
+        private void ApplyApplicationError(BsonDocument applicationError)
+        {
+            var expectedKeys = new[]
+            {
+                "address",
+                "generation", // optional
+                "maxWireVersion",
+                "when",
+                "type",
+                "response" // optional
+            };
+            JsonDrivenHelper.EnsureAllFieldsAreValid(applicationError, expectedKeys);
+            var address = applicationError["address"].AsString;
+            var endPoint = EndPointHelper.Parse(address);
+            var server = (Server)_serverFactory.GetServer(endPoint);
+            var connectionId = new ConnectionId(server.ServerId);
+            var type = applicationError["type"].AsString;
+            var maxWireVersion = applicationError["maxWireVersion"].AsInt32;
+            Exception simulatedException = null;
+            switch (type)
+            {
+                case "command":
+                    var response = applicationError["response"].AsBsonDocument;
+                    var command = new BsonDocument("Link", "start!");
+                    simulatedException = ExceptionMapper.MapNotPrimaryOrNodeIsRecovering(connectionId, command, response, "errmsg");
```

Comment:
We use `ExceptionMapper.MapNotPrimaryOrNodeIsRecovering` because that's what's called in [`ProcessResponse`](https://github.com/vincentkam/mongo-csharp-driver/blob/dec7cb4e2bdf5eb1646d76d1e3cf27e550f696f8/src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs#L442)
```
                if (rawDocument.GetValue("ok", false).ToBoolean())
                {
                    if (rawDocument.TryGetValue("recoveryToken", out var rawRecoveryToken))
                    {
                        var recoveryToken = ((RawBsonDocument)rawRecoveryToken).Materialize(binaryReaderSettings);
                        _session.CurrentTransaction.RecoveryToken = recoveryToken;
                    }
                }
                else
                {
                    var materializedDocument = rawDocument.Materialize(binaryReaderSettings);

                    var commandName = _command.GetElement(0).Name;
                    if (commandName == "$query")
                    {
                        commandName = _command["$query"].AsBsonDocument.GetElement(0).Name;
                    }

                    var notPrimaryOrNodeIsRecoveringException = ExceptionMapper.MapNotPrimaryOrNodeIsRecovering(connectionId, _command, materializedDocument, "errmsg");
                    if (notPrimaryOrNodeIsRecoveringException != null)
                    {
                        throw notPrimaryOrNodeIsRecoveringException;
                    }

```